examples/apps/windows/DatagramClientContext.cpp

```diff
@@ -0,0 +1,283 @@
+            // both the socket and object that serves its events have the same lifetime.
+            CoreApplication::Properties->Insert("clientContext", this);
+        }
+        catch (Exception^ ex)
+        {
+            _notify->NotifyFromAsyncThread(
+                "Remove clientContext error: " + ex->Message,
+                NotifyType::Error);
+        }
+        catch (task_canceled&)
+        {
+        }
+    }).then([this]()
+    {
+        auto endpointPair = ref new EndpointPair(_args->ClientHostName, _args->ClientPort,
+            _args->ServerHostName, _args->ServerPort);
+
+        _notify->NotifyFromAsyncThread("Start connecting", NotifyType::Status);
+
+        create_task(_client->ConnectAsync(endpointPair)).then(
+            [this, endpointPair](task<void> prevTask)
+        {
+            try
+            {
+                // Try getting an exception.
+                prevTask.get();
+                _notify->NotifyFromAsyncThread(
+                    "Connect from " + endpointPair->LocalHostName->CanonicalName +
+                    " to " + endpointPair->RemoteHostName->CanonicalName,
+                    NotifyType::Status);
+                SetConnected(true);
+            }
+            catch (Exception^ ex)
+            {
+                CoreApplication::Properties->Remove("clientContext");
+                _notify->NotifyFromAsyncThread(
+                    "Start binding failed with error: " + ex->Message,
+                    NotifyType::Error);
+            }
+            catch (task_canceled&)
+            {
+                CoreApplication::Properties->Remove("clientContext");
+            }
+        });
+    });
+}
+
+void
+DatagramClientContext::Send_Click(
+    Object^          sender,
+    RoutedEventArgs^ e,
+    String^          input)
+{
+    SendMessage(GetDataWriter(), input);
+}
+
+IAsyncAction^
+DatagramClientContext::CancelIO()
+{
+    return _client->CancelIOAsync();
+}
+
+void
+DatagramClientContext::SetConnected(
+    bool connected)
+{
+    _connected = connected;
+}
+
+bool
+Thread::DatagramClientContext::IsConnected() const
+{
+    return _connected;
+}
+
+void
+Thread::DatagramClientContext::OnMessage(
+    DatagramSocket^           socket,
+    MessageReceivedEventArgs^ eventArgs)
+{
+    try
+    {
+        auto dataReader = eventArgs->GetDataReader();
+        Receive(dataReader, dataReader->UnconsumedBufferLength);
+    }
+    catch (Exception^ ex)
+    {
+        auto socketError = SocketError::GetStatus(ex->HResult);
+        if (socketError == SocketErrorStatus::ConnectionResetByPeer)
+        {
+            // This error would indicate that a previous send operation resulted in an ICMP "Port Unreachable" message.
+            _notify->NotifyFromAsyncThread(
+                "Peer does not listen on the specific port. Please make sure that you run step 1 first " +
+                "or you have a server properly working on a remote server.",
```

Comment:
I dont know if the windows compiler will optimize out this string addition in its current form, but you could just take out the `+` sign and the compiler should take care of the string concatenation.